### ISSUE TITLE
[Question] Logical Issue in Upper-Level Model Design — Incorrect Connection Between Sample Weights and Loss

### ISSUE THREAD

**Issue by YiJin1107** [callous-youth/BOAT#64]
### Required prerequisites

- [x] I have read the documentation <https://boat.readthedocs.io/en/main/>.
- [x] I have searched the [Issue Tracker](https://github.com/callous-youth/BOAT/issues) to make sure this question has not already been answered.

### Problem Description

In `data_hyper_cleaning.py`, the `Net_x` class is currently implemented as follows:

```python
class Net_x(torch.nn.Module):    
    def __init__(self, tr):       
        super(Net_x, self).__init__()        
        self.x = torch.nn.Parameter(            
            torch.zeros(tr.data.shape[0]).to(device).requires_grad_(True)        
        )   
 
    def forward(self, y):        
        y = torch.sigmoid(self.x) * y        
        y = y.mean()        
        return y
```

### Issue

There appears to be a logical issue in how the sample weights are connected to the loss.

According to bilevel optimization theory, the sample weight should be applied to the loss of each individual sample, for example:

```python
weighted_loss = (sigma(theta_i) * loss_i).mean()
```

However, in the current implementation, `y` seems to already represent the averaged loss over all samples. Multiplying it by `torch.sigmoid(self.x)` and then taking the mean is effectively equivalent to scaling the overall loss globally by the average sample weight.

In other words, the current implementation does not perform differentiated weighting for each sample. Instead, it applies a global scaling factor to the whole loss.

### Expected Behavior

Each sample weight should be applied to the corresponding per-sample loss value before the final reduction operation.

The expected logic should be closer to:

```python
weighted_loss = (sample_weights * per_sample_losses).mean()
```

where:

- `sample_weights = torch.sigmoid(theta)`
- `per_sample_losses` contains the unreduced loss value for each training sample

### Suggested Fix

1. Apply the sample weights to the per-sample components of the lower-level loss, rather than to the already averaged lower-level loss.

2. Consider modifying the lower-level loss function so that it can return or receive a per-sample loss vector, allowing element-wise multiplication with the sample weight vector.

For example, the upper-level weighting module could be modified as follows:

```python
class Net_x(torch.nn.Module):    
    def __init__(self, tr):        
        super(Net_x, self).__init__()        
        self.weights = torch.nn.Parameter(            
            torch.zeros(tr.data.shape[0]).to(device).requires_grad_(True)        
        )    

    def get_sample_weights(self):        
        return torch.sigmoid(self.weights)
```

Then, in the lower-level training loss computation, the weighted loss could be computed as:

```python
sample_weights = net_x.get_sample_weights()
per_sample_losses = loss_fn(outputs, targets, reduction="none")
weighted_loss = (sample_weights * per_sample_losses).mean()
```

### Summary

The current implementation may not correctly realize sample-wise weighting. To better align with bilevel optimization, the sample weights should be applied before reducing the per-sample losses into a scalar.
